### PR TITLE
apimanagement : refactoring to use hashicorp/go-azure-sdk

### PR DESCRIPTION
**Pull request by sinbai**
Test results: There are 8 test cases failed in [this test](https://hashicorp.teamcity.com/buildConfiguration/TF_AzureRM_AZURERM_SERVICE_PUBLIC_APIMANAGEMENT/8397?buildTab=tests&status=failed). However, all of them have passed after re-running. See below for details:

https://hashicorp.teamcity.com/buildConfiguration/TF_AzureRM_AZURERM_SERVICE_PUBLIC_APIMANAGEMENT/9967?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildTestsSection=true&expandBuildChangesSection=true

https://hashicorp.teamcity.com/buildConfiguration/TF_AzureRM_AZURERM_SERVICE_PUBLIC_APIMANAGEMENT/10094?buildTab=tests

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
@sinbai heads up I'm looking into the base layer issues the tests have picked up here:

```
Error: creating/updating Api (Subscription: "*******"
        Resource Group Name: "acctestRG-230803055317928116"
        Service Name: "acctestAM-230803055317928116"
        Api: "acctestAMA-230803055317928116"): performing CreateOrUpdate: no applicable pollers were found for the response
```

**Comment by sinbai** [comment]
> 


> @sinbai heads up I'm looking into the base layer issues the tests have picked up here:
> 
> ```
> Error: creating/updating Api (Subscription: "*******"
>         Resource Group Name: "acctestRG-230803055317928116"
>         Service Name: "acctestAM-230803055317928116"
>         Api: "acctestAMA-230803055317928116"): performing CreateOrUpdate: no applicable pollers were found for the response
> ```


Hi @tombuildsstuff thanks for letting me know about this. After updating [ this line](https://github.com/hashicorp/go-azure-sdk/blob/0526c2a194a40a1f27cc3c023a5af5c9e5a7fad7/sdk/client/resourcemanager/poller.go#L43C24-L43C24) to the following code，the test case `TestAccApiManagementApiSchema_basic` passes on my local. Just letting you know, hope it helps.
 
`statusCodesToCheckProvisioningState := response.StatusCode == http.StatusOK || response.StatusCode == http.StatusCreated  || (lroIsSelfReference && response.StatusCode == http.StatusAccepted)`

**Comment by tombuildsstuff** [comment]
@sinbai indeed, that's the same fix I've just pushed a commit to fix :)

**Comment by tombuildsstuff** [comment]
Dependent on https://github.com/hashicorp/terraform-provider-azurerm/pull/22803

**Comment by tombuildsstuff** [comment]
@sinbai since the fix for this has been merged into `hashicorp/go-azure-sdk`, can you rebase this PR on top of `main`?

**Comment by sinbai** [comment]
> @sinbai since the fix for this has been merged into `hashicorp/go-azure-sdk`, can you rebase this PR on top of `main`?

Certainly, done:).

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks @sinbai

Looks like there are quite a few new test failures in TC though
image

File: internal/services/apimanagement/api_management_resource.go [hunk 1]
```diff
@@ -7,32 +7,40 @@ import (
 	"context"
 	"fmt"
 	"log"
+	"net/http"
 	"strconv"
 	"strings"
 	"time"
 
-	"github.com/Azure/azure-sdk-for-go/services/apimanagement/mgmt/2021-08-01/apimanagement" // nolint: staticcheck
+	"github.com/Azure/go-autorest/autorest"
```
Comment (tombuildsstuff) [inline internal/services/apimanagement/api_management_resource.go:15]: `hashicorp/go-azure-sdk` no longer uses `Azure/go-autorest` for APIManagement - so we can't rely on casting errors using it - can we just check the Status Codes being returned instead?

Comment (sinbai) [reply internal/services/apimanagement/api_management_resource.go]: @tombuildsstuff thanks for your feedback:).

**Comment by sinbai** [comment]
> Thanks @sinbai
> 
> Looks like there are quite a few new test failures in TC though image

@stephybun thanks for your feedback. I have fixed some issues and the code has been updated. Also, I started a new test on Teamcity as follows : https://hashicorp.teamcity.com/buildConfiguration/TF_AzureRM_AZURERM_SERVICE_PUBLIC_APIMANAGEMENT/8397?hideProblemsFromDependencies=false&hideTestsFromDependencies=false.

**Comment by sinbai** [comment]
> Thanks @sinbai
> 
> Looks like there are quite a few new test failures in TC though image

Hi @stephybun , given the test results in the PR comments, I assume this PR is ready for review, what do you think?

File: internal/services/apimanagement/api_management_api_data_source.go [hunk 2]
```diff
@@ -129,50 +129,48 @@ func dataSourceApiManagementApiRead(d *pluginsdk.ResourceData, meta interface{})
...
-		d.Set("version", props.APIVersion)
-		d.Set("version_set_id", props.APIVersionSetID)
-
-		if err := d.Set("protocols", flattenApiManagementApiDataSourceProtocols(props.Protocols)); err != nil {
+	name := getApiName(id.ApiId)
+	d.Set("name", name)
+	d.Set("resource_group_name", id.ResourceGroupName)
+
+	if model := resp.Model; model != nil && model.Properties != nil {
+		d.Set("description", pointer.From(model.Properties.Description))
```
Comment (stephybun) [inline internal/services/apimanagement/api_management_api_data_source.go:151]: For readability we assign `Properties` to a variable as well 
```suggestion
	if model := resp.Model; model != nil {
		if props := model.Properties; props != nil {
			d.Set("description", pointer.From(props.Description))
			...
		}
		
```

File: internal/services/apimanagement/api_management_api_diagnostic_resource.go [hunk 3]
```diff
@@ -264,55 +264,59 @@ func resourceApiManagementApiDiagnosticRead(d *pluginsdk.ResourceData, meta inte
...
-		if backend := props.Backend; backend != nil {
-			d.Set("backend_request", flattenApiManagementApiDiagnosticHTTPMessageDiagnostic(backend.Request))
-			d.Set("backend_response", flattenApiManagementApiDiagnosticHTTPMessageDiagnostic(backend.Response))
-		} else {
-			d.Set("backend_request", nil)
-			d.Set("backend_response", nil)
-		}
+	if model := resp.Model; model != nil {
+		d.Set("identifier", pointer.From(model.Name))
+		if model.Properties != nil {
```
Comment (stephybun) [inline internal/services/apimanagement/api_management_api_diagnostic_resource.go:290]:
```suggestion
		if props := model.Properties; props != nil {
```

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks @sinbai. I have one comment around assigning the response Model and Properties to variables within the read. We do this for readability, so that we remain consistent could you please update those in all the read methods of the resources/data sources that have been migrated? (I've only pointed out two in-line)

File: internal/services/apimanagement/api_management_api_data_source.go (hunk 2 above)
Comment (sinbai) [reply internal/services/apimanagement/api_management_api_data_source.go]: Fixed.

File: internal/services/apimanagement/api_management_api_diagnostic_resource.go (hunk 3 above)
Comment (sinbai) [reply internal/services/apimanagement/api_management_api_diagnostic_resource.go]: Fixed.

**Comment by sinbai** [comment]
> Thanks @sinbai. I have one comment around assigning the response Model and Properties to variables within the read. We do this for readability, so that we remain consistent could you please update those in all the read methods of the resources/data sources that have been migrated? (I've only pointed out two in-line)

@stephybun thanks for your feedback. Code has been updated. Could you please take another look?

**Review by stephybun — APPROVED** [review]
Thanks for making those changes to bring this in line with the other resources @sinbai. LGTM 🥟

**Pull request merged**